Don't wrap continuations from mnesia_ext backends

richcarl opened the pull request:
Continuations returned from mnesia_ext backends were originally wrapped with
the backend module, but that was a remnant of the earliest implementation,
and is not actually needed, since Mnesia already knows what the table
storage type is when it applies the continuation. When mnesia_ext was ported
to OTP 19, the wrapper got changed to use the table type alias instead of
the module, which triggered an error in the mnesia_eleveldb backend. This
patch removes the wrapping completely.

dgud commented:
Have I completely misunderstood what alias is used for?

As I understand it a backend module can implement several aliasas, so which alias is used need to known, in select/1 and repair_continuation.

The alias is sent to all other functions.

richcarl commented:
I also noted that the alias is included as a parameter in the other calls. I assume that the reasoning is that if a backend wants to include that information, it can simply encode it in the (opaque) continuation. Mnesia in itself never knows anything about what a continuation "is". But yes, that wrapper was there because in the earliest versions of mnesia_ext (I dug around in old git branches), the table type didn't encode what module to call.

dgud commented:
Hmm that can be tagged inside the Continuation directly.

dgud commented:
The tests fail, so you will need to update the ext_test module.

Is this really needed?
It will introduce a incompatibility in erlang/otp-20.
Though the only users might be you?

richcarl commented:
Updated repair_continuation/2 in ext_test. Didn't know you had added basic testing - that's a good thing, I guess. :-)

Yes, I think it's needed. It's a confusing thing that makes backends have to check whether an incoming continuation is wrapped or not, and in that case unwrap it. The mnesia_eleveldb backend did this, but only if it was tagged with the module name, not the alias. :-P - I have now updated it to handle all sorts of wrapped and non-wrapped continuations, but it sucks, and forcing other backends to do similar ugly stuff would be bad. Right now, mnesia_eleveldb is the only existing backend, so there should be no incompatibility issue.

dgud commented:
Yeah I had to write some tests, they are still very rudimentary.

PS: That is also why I had some (too) snarky comments about the supplied docs,
they didn't help me writing the plugin module, only half of the functions where there and
some where in the docs and not used in plugin and the number of args differed between the
docs and in the code. So I had too spend way to much time to add the PR to mnesia.
